File: backend/src/main/java/com/project/agent/service/ProductService.java
package com.project.agent.service;

import com.project.agent.domain.Product;
import com.project.agent.dto.ProductDTO;
import com.project.agent.exceptions.EntityAlreadyExistsException;
import com.project.agent.exceptions.EntityDoesNotExistException;
import com.project.agent.exceptions.InvalidDataException;

import java.util.ArrayList;

public interface ProductService {

    ArrayList<Product> getAllProducts();
    ArrayList<Product> getAgentProducts(String agentUsername) throws EntityDoesNotExistException;
    void createProduct(ProductDTO productDTO) throws EntityAlreadyExistsException, InvalidDataException, EntityDoesNotExistException;
    void updateProduct(String productId, ProductDTO productDTO) throws EntityDoesNotExistException, EntityAlreadyExistsException, InvalidDataException;
    void deleteProduct(String productId) throws EntityDoesNotExistException;
}

File: frontend/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';

@Injectable({
  providedIn: 'root'
})
export class ProductService {

  private readonly basePath = 'http://localhost:8081/product'

  constructor(private http: HttpClient) {
    this.http = http;
  }

  getAllProducts(){
    return this.http.get(this.basePath + '/get_all', {responseType: 'json'});
  }

  getAgentProducts(agentUsername : String){
    return this.http.get(this.basePath + '/' + agentUsername, {responseType: 'json'});
  }

  deleteProduct(productId) {
    var headers: HttpHeaders = new HttpHeaders({ 'Content-Type': 'application/json' });

    return this.http.post(this.basePath + '/' + productId + '/delete',
    {headers, responseType : 'text' as 'json'});
  }

  updateProduct(productId, updatedProduct) {
    var headers: HttpHeaders = new HttpHeaders({ 'Content-Type': 'application/json' });

    return this.http.post(this.basePath + '/' + productId + '/update', JSON.stringify(updatedProduct),
    {headers, responseType : 'text' as 'json'});
  }

  createProduct(newProduct) {
    var headers: HttpHeaders = new HttpHeaders({ 'Content-Type': 'application/json' });

    return this.http.post(this.basePath + '/create', JSON.stringify(newProduct),
    {headers, responseType : 'text' as 'json'});
  }
}
